python/tests/test_topology.py

```diff
@@ -134,40 +134,130 @@ def kc_distance(tree1, tree2, lambda_=0):
         raise ValueError("Trees must have the same samples")
     if not len(tree1.roots) == len(tree2.roots) == 1:
         raise ValueError("Trees must have one root")
-    sample_index_map = np.zeros(tree1.tree_sequence.num_nodes, dtype=int) - 1
-    for j, u in enumerate(samples):
-        sample_index_map[u] = j
+    for u in samples:
         if not tree1.is_leaf(u) or not tree2.is_leaf(u):
             raise ValueError("Internal samples not supported")
 
-    k = samples.shape[0]
-    n = (k * (k - 1)) // 2
-    m = [np.ones(n + k), np.ones(n + k)]
-    M = [np.zeros(n + k), np.zeros(n + k)]
+    n = samples.shape[0]
+    N = (n * (n - 1)) // 2
+    m = [np.zeros(N + n), np.zeros(N + n)]
+    M = [np.zeros(N + n), np.zeros(N + n)]
     for tree_index, tree in enumerate([tree1, tree2]):
-        stack = [(tree.root, 0, tree.time(tree.root))]
-        while len(stack) > 0:
-            node, depth, time = stack.pop()
-            children = tree.children(node)
-            for child in children:
-                stack.append((child, depth + 1, tree.time(child)))
-            for c1, c2 in itertools.combinations(children, 2):
-                for v1 in tree.samples(c1):
-                    index1 = sample_index_map[v1]
-                    for v2 in tree.samples(c2):
-                        index2 = sample_index_map[v2]
-                        a = min(index1, index2)
-                        b = max(index1, index2)
-                        pair_index = a * (a - 2 * k + 1) // -2 + b - a - 1
-                        assert m[tree_index][pair_index] == 1
-                        m[tree_index][pair_index] = depth
-                        M[tree_index][pair_index] = tree.time(tree.root) - time
-            if len(tree.children(node)) == 0:
-                index = sample_index_map[node]
-                M[tree_index][index + n] = tree.branch_length(node)
+        for n1 in range(n):
```

Comment:
If we're using this as the naive implementation, then let's use itertools to generate the sample pairs. We want two implementations:
1) naive_kc_distance which is the simplest, most Pythonic and readable code that we can come up with to compute the statistic
2) kc_distance which is mirror of the C implementation. (Maybe we should call this ``c_kc_distance`` for clarity?)

Reply:
Yup, I like that.